Adding other KAN layers

armbrusl opened the pull request:
I have implemented the following two layers 

Julia implementation of [FourierKAN](https://github.com/GistNoesis/FourierKAN)
Julia implementation of [ChebyKAN](https://github.com/SynodicMonth/ChebyKAN)

I also added another test for all the layers. Not sure how most of GitHub works so let me know if there is anything else I should do.

vpuri3 commented on .gitignore:
```diff
@@ -1 +1,2 @@
 Manifest.toml
+*.png
```
This should be in your personal global `.gitignore, not here.

vpuri3 replied:
revert this change.

vpuri3 commented on .vscode/settings.json:
```diff
@@ -0,0 +1,3 @@
+{
+    "julia.environmentPath": "/home/b4c/Documents/code/KolmogorovArnold.jl/test"
+}
```
This file does not belong here. You should likely ignore `.vscode/` in the global `.gitignore` on your machine.

vpuri3 replied:
do a git rm on this file

vpuri3 commented on Project.toml:
```diff
@@ -11,9 +11,16 @@ LinearAlgebra = "37e2e46d-f89d-539d-b4ee-838fcccc9c8e"
 NNlib = "872c559c-99b0-510c-b3b7-b6c96a88d5cd"
 Random = "9a3f8284-a2c9-5f02-9a11-845980a1fd5c"
+TensorOperations = "6aa20fa7-93e2-5fca-9bc0-fbd0db3c71a2"
+Test = "8dfed614-e22c-5e08-85e1-65c5234f0b40"
```
Why is Test a dependency here?

vpuri3 commented:
Thanks @armbrusl, the code looks ok. Can you 
- take a look at the comments
- update the README code to include the new layers

armbrusl commented on Project.toml:
```diff
@@ -11,9 +11,16 @@ LinearAlgebra = "37e2e46d-f89d-539d-b4ee-838fcccc9c8e"
+TensorOperations = "6aa20fa7-93e2-5fca-9bc0-fbd0db3c71a2"
+Test = "8dfed614-e22c-5e08-85e1-65c5234f0b40"
 WeightInitializers = "d49dbf32-c5c2-4618-8acc-27bb2598ef2d"
+cuTENSOR = "011b41b2-24ef-40a8-b3eb-fa098493e9e1"
```
Due to the batched_mul function, the ChebyKAN layer will not work on the GPU without this. Is it common practice to leave this up to the user to download ?

vpuri3 replied:
Yes. You don't want to install cuTENSOR for users who aren't running on the GPU. `TensorOperations` should throw a warning if a user is calling `@tensor` with CuArrays without loading `cuTENSOR`. Can you check if TensorOperations does that? If not, you can include a warning in `batched_mul`.

armbrusl replied:
I does throw a warning

vpuri3 replied:
Great, then we don't need to do anything

vpuri3 replied:
remove cuTENSOR from Project.toml and move it to test/Project.toml

vpuri3 commented:
just pushed a commit to master with all the Lux 0.x --> 1.x fixes. @armbrusl I also rebased this branch to master. I removed all the unnecessary packages in `test/Project.toml` (e.g., LLVM, Debugger, etc). They should not be added to the pull request. You should do a `git fetch && git pull` on your machine before proceeding. You should also update your Julia environment.

Can you also move the test you added in `test/runtests.jl` to a separate file.

vpuri3 commented on src/cdense.jl:
```diff
@@ -0,0 +1,67 @@
+
+    y = if addbias
+        y .+ p.B
+    end
```
same here

armbrusl commented:
I think I have addressed your suggestions. Thanks for walking me through it! Let me know if there is anything else I need to do. The only thing I haven't done is edit the README file, is there anything specific I should add ?

vpuri3 commented:
see above comments

vpuri3 commented on src/utils.jl:
```diff
@@ -60,4 +62,15 @@ function CRC.rrule(::typeof(_iqf), x)
+# Batched Multiplication using Einstein Summation
+#======================================================#
+
+@inline function batched_mul(x, coeffs)
```
Can you not use NNlib.batched_mul or NNlib.batched_vec for this?

vpuri3 replied:
https://fluxml.ai/NNlib.jl/dev/reference/#NNlib.batched_mul

armbrusl replied:
I am not sure. I have been trying but I have not able to get it right

vpuri3 commented:
Can you add the new layers to the code in the README and report timings?

https://github.com/vpuri3/KolmogorovArnold.jl/blob/master/README.md?plain=1#L14-L88

If not, I can do it after this is merged.

vpuri3 commented:
See workflow error: https://github.com/vpuri3/KolmogorovArnold.jl/actions/runs/11579597184/job/32236568354?pr=3#step:6:545

vpuri3 commented:
@armbrusl can you investigate this error? I am not seeing it in master with 1.10 or 1.11. I also updated master to only test on 1.10 and 1.11. If you rebase this branch, then that change will be incorporated.

armbrusl commented:
@vpuri3 I am not able to replicate the errors, when I run the required command 

import Pkg; Pkg.test(;coverage=true, julia_args=["--check-bounds=yes", "--compiled-modules=yes"], force_latest_compatible_version=false, allow_reresolve=true)

everything works fine. I also rebased the branch.

vpuri3 commented:
@armbrusl i tested the CI on Master in #4 and it is fine. There's some change introduced in this PR that is causing the failures.

I also don't see your rebase commits. Make sure you git checkout master, git fetch and git pull the latest version. then checkout this branch again and git merge master. and then do a git push

armbrusl commented:
I think I have done the steps you told me, how will I know if I have done it right

vpuri3 commented:
I'll check out this branch and investigate this evening

vpuri3 commented:
@armbrusl you haven't rebased correctly. Look at [this branch](https://github.com/armbrusl/KolmogorovArnold.jl/tree/master).

![image](https://github.com/user-attachments/assets/762733ab-669b-4a0f-809e-ab7549a499b0)
